components/skus/browser/skus_service.cc

```diff
@@ -0,0 +1,152 @@
+// Copyright (c) 2021 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// you can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "brave/components/skus/browser/skus_service.h"
+
+#include <utility>
+
+#include "base/json/json_reader.h"
+#include "base/task/post_task.h"
+#include "base/threading/sequenced_task_runner_handle.h"
+#include "brave/components/skus/browser/pref_names.h"
+#include "brave/components/skus/browser/rs/cxx/src/lib.rs.h"
+#include "brave/components/skus/browser/rs/cxx/src/shim.h"
+#include "brave/components/skus/browser/skus_context_impl.h"
+#include "brave/components/skus/browser/skus_utils.h"
+#include "components/prefs/pref_service.h"
+#include "services/network/public/cpp/shared_url_loader_factory.h"
+
+namespace {
+
+void OnRefreshOrder(skus::RefreshOrderCallbackState* callback_state,
+                    skus::SkusResult result,
+                    rust::cxxbridge1::Str order) {
+  std::string order_str = static_cast<std::string>(order);
+  if (callback_state->cb) {
+    std::move(callback_state->cb).Run(order_str);
+  }
+  delete callback_state;
+}
+
+void OnFetchOrderCredentials(
+    skus::FetchOrderCredentialsCallbackState* callback_state,
+    skus::SkusResult result) {
+  if (callback_state->cb) {
+    std::move(callback_state->cb).Run("");
+  }
+  delete callback_state;
+}
+
+void OnPrepareCredentialsPresentation(
+    skus::PrepareCredentialsPresentationCallbackState* callback_state,
+    skus::SkusResult result,
+    rust::cxxbridge1::Str presentation) {
+  if (callback_state->cb) {
+    std::move(callback_state->cb).Run(static_cast<std::string>(presentation));
+  }
+  delete callback_state;
+}
+
+void OnCredentialSummary(skus::CredentialSummaryCallbackState* callback_state,
+                         skus::SkusResult result,
+                         rust::cxxbridge1::Str summary) {
+  std::string summary_string = static_cast<std::string>(summary);
+  base::JSONReader::ValueWithError value_with_error =
+      base::JSONReader::ReadAndReturnValueWithError(
+          summary_string, base::JSONParserOptions::JSON_PARSE_RFC);
+  absl::optional<base::Value>& records_v = value_with_error.value;
+
+  if (records_v && callback_state->prefs) {
+    if (callback_state->domain == "vpn.brave.com" ||
+        callback_state->domain == "vpn.bravesoftware.com" ||
+        callback_state->domain == "vpn.brave.software") {
+      const base::Value* active = records_v->FindKey("active");
+      if (active) {
+        bool has_credential = active && active->is_bool() && active->GetBool();
+        callback_state->prefs->SetBoolean(skus::prefs::kSkusVPNHasCredential,
```

Comment:
I'm not sure exactly what `kSkusVPNHasCredential` is (see comment below about explaining these, but I'm concerned that we're storing information about a transient state in persistent storage.

Reply:
The intention is to store whether the profile has ever had a credential. If so, we can call `credential_summary` at any time to get a status. If a credential is present, the response will indicate that. It should note whether there was a credential but it's expired. If it's false, the user has never tried to purchase VPN.

Reply:
Comment added to the original preference 👍